File: VB/RangedMaskSample/EditElementCollection.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections

Namespace RangedMaskSample
	Public Class EditElementCollection
		Inherits CollectionBase
		Public Sub New()

		End Sub
		Public Sub New(ByVal capacity As Integer)
			MyBase.New(capacity)

		End Sub

		Public Overridable Sub AddRange(ByVal elements() As EditElement)
			For Each element As EditElement In elements
				List.Add(element)
			Next element
		End Sub

		Public Overridable Function Add(ByVal element As EditElement) As Integer
			Dim res As Integer = IndexOf(element)
			If res = -1 Then
				res = List.Add(element)
			End If
			Return res
		End Function

		Public Function IndexOf(ByVal element As EditElement) As Integer
			Return List.IndexOf(element)
		End Function

		Default Public Property Item(ByVal index As Integer) As EditElement
			Get
				Return (CType(List(index), EditElement))
			End Get
			Set(ByVal value As EditElement)
				List(index) = value
			End Set
		End Property

		Public Sub Insert(ByVal index As Integer, ByVal value As EditElement)
			List.Insert(index, value)
		End Sub

		Public Sub Remove(ByVal aEdit As MaskButtonEdit)
			Dim aElement As EditElement = Find(aEdit)
			If aElement IsNot Nothing Then
				List.Remove(aElement)
			End If
		End Sub

		Public Function Contains(ByVal aEdit As MaskButtonEdit) As Boolean
			Return Find(aEdit) IsNot Nothing
		End Function

		Public Function Find(ByVal aEdit As MaskButtonEdit) As EditElement
			For i As Integer = 0 To Count - 1
				If Me(i).OwnerEdit Is aEdit Then
					Return Me(i)
				End If
			Next i
			Return Nothing
		End Function
	End Class
End Namespace

File: CS/RangedMaskSample/EditElementCollection.cs
using System;
using System.Collections;

namespace RangedMaskSample
{
    public class EditElementCollection : CollectionBase
    {
        public EditElementCollection()
        {

        }
        public EditElementCollection(int capacity)
            : base(capacity)
        {

        }

        public virtual void AddRange(EditElement[] elements)
        {
            foreach (EditElement element in elements)
            {
                List.Add(element);
            }
        }

        public virtual int Add(EditElement element)
        {
            int res = IndexOf(element);
            if (res == -1) res = List.Add(element);
            return res;
        }

        public int IndexOf(EditElement element)
        {
            return List.IndexOf(element);
        }

        public EditElement this[int index]
        {
            get
            {
                return ((EditElement)List[index]);
            }
            set
            {
                List[index] = value;
            }
        }

        public void Insert(int index, EditElement value)
        {
            List.Insert(index, value);
        }

        public void Remove(MaskButtonEdit aEdit)
        {
            EditElement aElement = Find(aEdit);
            if (aElement != null)
                List.Remove(aElement);
        }

        public bool Contains(MaskButtonEdit aEdit)
        {
            return Find(aEdit) != null;
        }

        public EditElement Find(MaskButtonEdit aEdit)
        {
            for (int i = 0; i < Count; i++)
            {
                if (this[i].OwnerEdit == aEdit)
                {
                    return this[i];
                }
            }
            return null;
        }
    }
}
